tests/fmt_spec.lua

```diff
@@ -0,0 +1,46 @@
+local replace_placeholders = require('config.plugins.luasnip.utils').replace_placeholders
+
+local check = function(msg, fmt, args, expected)
+    it(msg, function()
+        local result = replace_placeholders(fmt, args)
+        assert.are.same(expected, table.concat(result))
+    end)
+end
+
+describe('replace_placeholders', function()
+    check('expands with no numbers',
+        'a{}b{}c{}d', {4, 5, 6},
+        'a4b5c6d')
+
+    check('expands with explicit numbers',
+        'a{2}b{1}c{3}d', {4, 5, 6},
+        'a5b4c6d')
+
+    check('expands with mixed numbering',
+        'a{}b{3}c{}d', {1, 2, 3, 4},
+        'a1b3c4d')
+
+    check('expands named placeholders',
+        'a{A}b{B}c{C}d', {A = 1, B = 2, C = 3},
+        'a1b2c3d')
+
+    check('expands all mixed',
+        'a {A} b {} c {3} d {} e {B} f {A} g', {1, 2, 3, 4, A = 10, B = 20},
+        'a 10 b 1 c 3 d 4 e 20 f 10 g')
+
+    check('excludes trailing text',
+        '{}abcd{}', {1, 2},
+        '1abcd2')
+
+    check('ignores {{}}',
+        'a{{}}b{}c{{}}d{}e', {2, 4},
+        'a{}b2c{}d4e')
+
+    check('ignores {{text}}',
+        'a{{d}}b{}c', {2},
+        'a{d}b2c')
```

Comment:
Why is it not `a{{d}}b2c` ?

How would you write a snippet to output: `foo {{bar}} baz` ? (using a lot of nested braces? using `\` to escape brances?)

Reply:
The double brackets are to escape single ones, eg `{{.*}}` - > `{.*}`, for literal double brackets you'd need `{{{{text}}}}`, but I'm not sure if that'll work right now.

Reply:
The syntax is based on [fmtlib](https://fmt.dev/latest/syntax.html) and [Python formatting](https://www.python.org/dev/peps/pep-3101/). I checked that it currently works only for the case of `{{...}}` -> `{...}`. Currently, for more braces it gives incorrect results. I would need to do some tests to modify the algorithm if we want to handle such cases.

Reply:
Maybe doing `%b{}` for balanced brackets is enough?

Reply:
Oh, and reading through pythons fmt, it seems like singular `{` can also be escaped with `{{`, which would make more sense to me then only being able to escape pairs.

Reply:
Didn't know about the `%b{}` pattern. But it doesn't seem to address the issue with escaping and we would need some extra logic to manually detect if this should be escaped. `%b{}` for a string like `a {{ b }} c` matches the whole `{{ b }}`, so we would need to run it recursively for that ` b ` part. 

Also, as the pattern currently doesn't accept spaces between braces so this should greatly reduce the number of scenarios in which the user has to do the escaping. It might be a bit confusing for someone that expects full compatibility with fmtlib/Python formatting but I feel like it might be more practical in this particular case.

Reply:
Oh, I thought the problem was that it matches
```txt
{{{{text}}}}
^       ^
```
these two together, nvm then.
Not sure if using a pattern to find matching brackets is the best approach(I don't know if it's possible to ignore escaped brackets with it). 
What would probaly work is building a table containing opening brackets as keys and closing ones as the respective values (eg. `aaa{b{e}c}dd` -> `{4 = 10, 6 = 8}`) (see [here](https://github.com/L3MON4D3/LuaSnip/blob/master/lua/luasnip/util/parser.lua#L24-L48), rules for escaping would need to be adjusted, not entirely sure if factoring out is worth it) and then splitting the text on these values.